packages/hash/api/src/graph/knowledge/system-types/file.ts

```diff
@@ -0,0 +1,181 @@
+
+import { EntityTypeMismatchError } from "../../../lib/error";
+import { PresignedPostUpload } from "../../../storage";
+import { genId } from "../../../util";
+import { ImpureGraphFunction, PureGraphFunction } from "../..";
+import { SYSTEM_TYPES } from "../../system-types";
+import { createEntity, CreateEntityParams } from "../primitive/entity";
+
+const MAX_FILE_SIZE_BYTES = 1000 * 1000 * 1000;
+
+const UPLOAD_URL_EXPIRATION_SECONDS = 60 * 30;
+
+export type FileKey =
+  | {
+      type: "ObjectStoreKey";
+      objectStoreKey: string;
+    }
+  | {
+      type: "ExternalFileLink";
+      externalFileLink: string;
+    };
+
+export type File = {
+  fileUrl: string;
+  fileMediaType: string;
+  fileKey: FileKey;
+};
+
+export const getFileFromEntity: PureGraphFunction<{ entity: Entity }, File> = ({
+  entity,
+}) => {
+  if (
+    entity.metadata.entityTypeId !== SYSTEM_TYPES.entityType.file.schema.$id
+  ) {
+    throw new EntityTypeMismatchError(
+      entity.metadata.editionId.baseId,
+      SYSTEM_TYPES.entityType.block.schema.$id,
+      entity.metadata.entityTypeId,
+    );
+  }
+
+  const fileUrl = entity.properties[
+    SYSTEM_TYPES.propertyType.fileUrl.metadata.editionId.baseId
+  ] as string;
+
+  const fileMediaType = entity.properties[
+    SYSTEM_TYPES.propertyType.fileMediaType.metadata.editionId.baseId
+  ] as string;
+
+  const fileKeyObject = entity.properties[
+    SYSTEM_TYPES.propertyType.fileKey.metadata.editionId.baseId
+  ] as Record<string, any>;
+
+  const fileKey: FileKey =
+    SYSTEM_TYPES.propertyType.externalFileLink.metadata.editionId.baseId in
+    fileKeyObject
+      ? {
+          type: "ExternalFileLink",
+          externalFileLink: fileKeyObject[
+            SYSTEM_TYPES.propertyType.externalFileLink.metadata.editionId.baseId
+          ] as string,
+        }
+      : {
+          type: "ObjectStoreKey",
+          objectStoreKey: fileKeyObject[
+            SYSTEM_TYPES.propertyType.objectStoreKey.metadata.editionId.baseId
+          ] as string,
+        };
+
+  return {
+    fileUrl,
+    fileMediaType,
+    fileKey,
+  };
+};
+
+export const formatUrl = (key: string) => {
+  return `${apiOrigin}/file/${key}`;
+};
+
+export const createFileFromUploadRequest: ImpureGraphFunction<
+  Omit<CreateEntityParams, "properties" | "entityTypeId"> & {
+    size: number;
+    mediaType: string;
+  },
+  Promise<{ presignedPost: PresignedPostUpload; entity: Entity }>
+> = async (
+  ctx,
+  params,
+): Promise<{ presignedPost: PresignedPostUpload; entity: Entity }> => {
+  const { uploadProvider } = ctx;
+  const { ownedById, actorId, mediaType, size } = params;
+
+  /**
+   * @todo we want the downstream upload services to enforce the given size
+   * limit, otherwise the size limits are meaningless
+   */
+  if (size > MAX_FILE_SIZE_BYTES) {
+    throw new Error("The file is heavier than the maximum allowed file size");
+  }
+
+  const fileIdentifier = genId();
```

Comment:
I did consider doing that, but since I can imagine a file will have it's underlying file changed when updated, i opted to generate a unique identifier instead. As the `entityUuid` would not be unique across entity editions, we would have a problem if we were to use the same name for the file editions in s3 (we'd overwrite the underlying file instead of creating new ones). I would need the entity edition timestamp or equivalent, and that is not possible to pre-populate currently.

Reply:
Understood, thanks for explaining :)